pandas/io/parquet.py

```diff
@@ -147,27 +148,30 @@ def to_parquet(df, path, engine='auto', compression='snappy', **kwargs):
 
     valid_types = {'string', 'unicode'}
 
-    # validate index
-    # --------------
-
     # validate that we have only a default index
     # raise on anything else as we don't serialize the index
-
-    if not isinstance(df.index, Int64Index):
-        raise ValueError("parquet does not support serializing {} "
-                         "for the index; you can .reset_index()"
-                         "to make the index into column(s)".format(
-                             type(df.index)))
-
-    if not df.index.equals(RangeIndex.from_range(range(len(df)))):
-        raise ValueError("parquet does not support serializing a "
-                         "non-default index for the index; you "
-                         "can .reset_index() to make the index "
-                         "into column(s)")
-
-    if df.index.name is not None:
-        raise ValueError("parquet does not serialize index meta-data on a "
-                         "default index")
+    # *unless* we're using pyarrow >= 0.7.1 which does support multi-indexes
```

Comment:
so it *might* be better (though a bit more refactoring), to move the validation of index and validation of columns into methods on a base class to the impl

e.g.

```
class BaseImpl(object):


    def validate_index(....):
       ..

    def validate_columns(....):
      ....


class PyArrowImpl(BaseImpl):

     def write(...):
          if pyarrow < 0.7.1:
             self.validate_index()

         self.validate_columns()

class FastParquetImpl(BaseImpl):
      .....
```

as it makes things cleaner.

Reply:
Refactored in the latest commits...